test/src/unit-class_parser.cpp

```diff
@@ -45,7 +45,8 @@ json parser_helper(const std::string& s)
     // if this line was reached, no exception ocurred
     // -> check if result is the same without exceptions
     json j_nothrow;
-    CHECK_NOTHROW(json::parser(nlohmann::detail::input_adapter(s), nullptr, false).parse(true, j_nothrow));
+    std::unique_ptr<json::exception> ex;
```

Comment:
What is the advantage of a `unique_ptr` over `std::exception_ptr`?

Reply:
1. `std::exception_ptr` isn't limited to `json::exception` types.
2. The only things you can do with an `std::exception_ptr` are convert to bool, compare to another pointer, and rethrow.  You can't actually get the exception object out of it, so all the caller would be able to do is determine whether or not an exception was thrown, not what that exception was.

Reply:
Right, there's no way to retrieve the exception, forgot about that.

I do think we should not use `std::unique_ptr<json::exception>` though, but `std::error_code` instead (this [blog post](https://akrzemi1.wordpress.com/2017/07/12/your-own-error-code/) is really useful on this topic).

`std::filesystem` uses `std::error_code` for the `noexcept` overloads, one can still embed a lot of information with this facility. I've seen it mentioned in [Outcome's documentation](https://ned14.github.io/outcome/tutorial/error_code) too.

What do you all think?

Reply:
I don't work without exceptions, so I don't really care one way or another.  I think the best solution in the end is probably to return an `std::expected` or `std::outcome` or whatever it ends up as, but that's years away.

Reply:
I agree that long term using `std::expected` or `std::outcome` would be best. My original plan was to use `std::error_code` (as you can see by my branch name) but it has it's own issues.  Mostly creating a custom formatted error message with context does not seem cleanly possible. You can see in the blog post linked (which I used as a source as well) that aspect is glossed over.

Reply:
Indeed you cannot have context in the `error_code` (e.g. line of the parsing failure).

However I don't think that's relevant, we are used to this already. Trying to copy an non-existing file with `std::filesystem` gives the same behavior:

exception:
> what: filesystem error: cannot copy: No such file or directory [asdfasd] [asdfasdfa2] 

error_code:
> No such file or directory

We will not have the line number or very precise information like that, but @nlohmann already documented all the exceptions thrown by the library, with a code and example message, which are IMO more than enough!

It should be quite easy to map each exception type to one enum/error category, I believe we should also add the `error_code` as a member of the exception.

Reply:
I guess I would argue that losing context on json parsing errors would be quite a step backwards. Whereas not a loss for the filesystem examples as the extra context can be inferred by the calling parameters.

Another option could be yet another overload taking in another parameter `std::string &error_location` or combine both std::error_code and std::string error_location into one object.

Reply:
Right now `std::abort` is called, so I don't think this is a step backward :p. If I understand correctly, one must use `dynamic_cast` to down cast the output parameter `unique_ptr<json::exception>`.

If so, that's really a big issue, usually people that turn off exceptions also turn off RTTI. Not even mentioning how painful this is to use, nor the cost of heap allocation each time the `unique_ptr` gets constructed.

I don't see how we could combine an `error_location` parameter with the `error_code`, the former being relevant only in a few cases (e.g. parse errors). Maybe a `extra_detail` field?
Even that would be sub-optimal, as the error type would bear the size of an extra string, even when it's not used.

On my machine it goes from 16 bytes (error_code) to 48 (error_code + string).

I would suggest trying to implement the `error_code` version to see how it goes, we might find the messages detailed enough!

Reply:
The thing is that the new overload does not call `std::abort` at all. It just passes the exception with the context back up to the caller.

No RTTI is required either. The virtual `what()` function dumps the context of the parse error with full context for the caller.

That is what I'm referring to as a step backwards.

Reply:
Sorry, I misread the latest commits.

I did not realize that the definition of `json::exception` depends on whether or not the exceptions are disabled globally. This is very misleading, if someone wants to use the nothrow version (`noexcept` should be added by the way) with exceptions enabled, they will not have the same `json::exception` struct than someone without exceptions. One should have the choice to use the throw or nothrow version in the code and not through compiler flags.

You could argue that this is not a big deal, since the data member is not exposed (it does change the size of the struct though). Plus, having a parameter of type `json::exception` which is not an exception feels strange.

What I consider a step backwards is the loss of the ability to switch on the error type.
With exceptions, it can still be done with multiple `catch`, or by downcasting.
With `std::error_code`, a simple `switch`, or an approach as discussed in [this paper](http://open-std.org/JTC1/SC22/WG21/docs/papers/2017/p0824r0.pdf) can be used.

Here, users cannot do anything, they have a `std::string`. They might parse it, but frankly that's unusable. The only advantage that I see is for debugging purposes, but it's not possible to take a decision at runtime based on this approach.

I would recommend the reading of the aforementioned paper, especially the part about interoperability with four libraries using `std::error_code`. Also, this mechanism is used in the latest Couroutine/Networking TS, adopting it seems future proof to me.

Let's not forget this library is likely to be used in/with other libraries:
For example, [Boost.Beast](https://github.com/boostorg/beast) uses `error_code` extensively, if someone wants to wrap it with this library (e.g. to parse the json body of an HTTP request), having both libraries using/providing `std::error_code` is quite helpful.

I'm far from being an expert in Beast, though @vinniefalco stated in [this issue](https://github.com/nlohmann/json/issues/774#issuecomment-339323015) that one could use this library to implement a Beast `Body` container (assuming incremental parsing is supported). I might be wrong but it seems that having `error_code` would simplify the implementation of `BodyReader/BodyWriter` concepts too.

Reply:
I definitely see where you are coming from @theodelrieu and pretty much agree with every one of your points. I appreciate the links.

The issue is in regards to a specific, but I would argue potentially popular, scenario though. That is using this library for parsing json and getting nice context aware error messages on failures even if exception handling is completely disabled. (most common in game engines, embedded)

Being able to handle various error codes / categories at runtime for parse errors is not important here, and I can't see the ability to do so being useful other than a theoretical exercise.

@nlohmann does the above description answer your question on why this PR exists?

Reply:
So basically all you want to have is access to parser diagnostics (error message + location) when exceptions are switched off?

Reply:
yes

Reply:
@jseward I understand the use-case, I think a compromise can be reached.

The good solution would be [outcome](https://github.com/ned14/outcome), you can indeed attach an extra payload to it. Unfortunately, I do not think we can use it in this library.

However, we could have the following type:

```cpp
// not a good name (std::system_error is used as an exception).
struct json_error {
  std::error_code ec;
  std::string context; // empty if ec == 0, or no context provided.
};
```

This type would replace the `std::unique_ptr<json::exception>` in this PR. Plus, the `#ifdef` inside `json::exception` definition can be removed.

I would also suggest adding an `error_code` to `json::exception` too (for the same reasons mentioned before in this thread, similarly to `std::filesystem_error`).

What do you all think?

Reply:
We already support parsing without exceptions (and return a `discarded` type on parse errors). I think the desired outcome could be achieved by just adding the required member functions to the parser class to query the required diagnostics in case an error occurs. The same functions could be re-used in the code that creates the actual `parse_error` exception so we would not even need to add too much code.